src/crimson/common/gated.h

```diff
@@ -42,14 +53,80 @@ class Gated {
     });
   }
 
+  template <typename Func>
+  auto simple_dispatch(const char* what, Func&& func) {
+    return seastar::with_gate(pending_dispatch, std::forward<Func>(func));
+  }
+
   seastar::future<> close() {
+    gated_logger().debug("closing gate on shard {} on shard {}", sid, seastar::this_shard_id());
```

Comment:
done

Reply:
I can't find the assertion in the new commit.

Reply:
@cyx1231st  asserting during close will cause some issues with io_handle, we are doing close of prev sid (cleanup_prv_shard) and trying to close the gate on a different shard it was created on. checking that

Reply:
> @cyx1231st do you have any opinion on io_handler, do you think it should be changed as well to use gate_per_shard?
it should run on each core, but currently using single gate

I think not, `io_handler` can only run on a single core (it is owned by `Connection` living on one core as well). However, it (as well as `Connection`) may switch cores due to connection recovery (which is a complicated mechanism), probably there were some unexpected issues revealed by adding `Gated::sid` asserts in this PR. But I believe it shouldn't use `gate_per_shard` for the reasons.

Reply:
See `IOHandler::cleanup_prv_shard(prv_sid)`:
https://github.com/ceph/ceph/blob/dc8d75f11fefba4d39f92288ea28398499202f82/src/crimson/net/io_handler.cc#L656-L664

Looks to me this part is correct, because the logic is submitted to `prv_sid` in calling `prv_states.close()`, and `prv_states` (containing the gate) is destructed inside `prv_sid` as well.

Reply:
However, it seems to be an issue when switch to a new core https://github.com/ceph/ceph/blob/dc8d75f11fefba4d39f92288ea28398499202f82/src/crimson/net/io_handler.cc#L734-L735

^ The new `shard_states` (and shard_states_t::gate) isn't created by the new shard `new_sid` where it will be used.

---

A potential fix might be introducing another constructor for `Gated` to specify sid where it will be used, so that:
https://github.com/ceph/ceph/blob/dc8d75f11fefba4d39f92288ea28398499202f82/src/crimson/net/io_handler.h#L257-L258
can be modified as:
```c++
shard_states_t(seastar::shard_id _sid, io_state_t state)
  : sid{_sid}, io_state{state}, gate{_sid} {}  // allow gate to be constructed by one sid, and used by another sid
```

If this looks reasonable, can you have a try to see whether it works?

Reply:
@cyx1231st that will work, thanks!